File: libsignal/service/src/main/java/org/whispersystems/signalservice/internal/registrationpin/PinValidityChecker.java
package org.whispersystems.signalservice.internal.registrationpin;

public final class PinValidityChecker {

  public static boolean valid(String pin) {
    pin = pin.trim();

    if (pin.isEmpty()) {
      return false;
    }

    if (PinString.allNumeric(pin)) {
      pin = PinString.toArabic(pin);

      return !sequential(pin) &&
             !sequential(reverse(pin)) &&
             !allTheSame(pin);
    } else {
      return true;
    }
  }

  private static String reverse(String string) {
    char[] chars = string.toCharArray();

    for (int i = 0; i < chars.length / 2; i++) {
      char temp = chars[i];
      chars[i] = chars[chars.length - i - 1];
      chars[chars.length - i - 1] = temp;
    }

    return new String(chars);
  }

  private static boolean sequential(String pin) {
    int length = pin.length();

    if (length == 0) {
      return false;
    }

    char c = pin.charAt(0);

    for (int i = 1; i < length; i++) {
      char n = pin.charAt(i);
      if (n != c + 1) {
        return false;
      }
      c = n;
    }

    return true;
  }

  private static boolean allTheSame(String pin) {
    int length = pin.length();

    if (length == 0) {
      return false;
    }

    char c = pin.charAt(0);

    for (int i = 1; i < length; i++) {
      char n = pin.charAt(i);
      if (n != c) {
        return false;
      }
    }

    return true;
  }
}

File: lib/network/src/main/java/org/signal/network/pin/PinValidityChecker.kt
/*
 * Copyright 2026 Signal Messenger, LLC
 * SPDX-License-Identifier: AGPL-3.0-only
 */

package org.signal.network.pin

import org.whispersystems.signalservice.api.kbs.PinString

/**
 * Rejects PINs that are trivially guessable. A numeric PIN must not be empty, sequential in either
 * direction, or a single repeated digit. Non-numeric PINs are only checked for emptiness.
 */
object PinValidityChecker {

  @JvmStatic
  fun valid(pin: String): Boolean {
    val trimmed = pin.trim()

    if (trimmed.isEmpty()) {
      return false
    }

    if (!PinString.allNumeric(trimmed)) {
      return true
    }

    val arabic = PinString.toArabic(trimmed)

    return !arabic.isSequential() && !arabic.reversed().isSequential() && !arabic.isSingleRepeatedChar()
  }

  private fun String.isSequential(): Boolean {
    return zipWithNext().all { (previous, next) -> next == previous + 1 }
  }

  private fun String.isSingleRepeatedChar(): Boolean {
    return all { it == this[0] }
  }
}
